File: src/lib/algorithms.ts
export interface BankersInput {
  processes: string[];
  resources: string[];
  allocation: number[][];
  max: number[][];
  available: number[];
}

export interface DeadlockInput {
  nodes: { id: string; type: 'process' | 'resource' }[];
  edges: { source: string; target: string }[];
}

export interface RecoveryInput {
  allocation: number[][];
  available: number[];
  deadlockedProcesses: string[];
  strategy: 'terminate-all' | 'terminate-one-by-one' | 'preemption';
}

/**
 * Banker's Algorithm implementation
 */
export function bankersSafetyCheck(input: BankersInput) {
  const { processes, resources, allocation, max, available } = input;
  const n = processes.length;
  const m = resources.length;

  const need = max.map((row, i) => row.map((val, j) => val - allocation[i][j]));
  const work = [...available];
  const finish = new Array(n).fill(false);
  const safeSequence: string[] = [];
  const log: string[] = [];

  log.push(`Initial Work: [${work.join(', ')}]`);

  let count = 0;
  while (count < n) {
    let found = false;
    for (let i = 0; i < n; i++) {
      if (!finish[i]) {
        let j;
        for (j = 0; j < m; j++) {
          if (need[i][j] > work[j]) break;
        }

        if (j === m) {
          log.push(`Process ${processes[i]} can be allocated resources. Need: [${need[i].join(', ')}] <= Work: [${work.join(', ')}]`);
          for (let k = 0; k < m; k++) {
            work[k] += allocation[i][k];
          }
          finish[i] = true;
          safeSequence.push(processes[i]);
          log.push(`Process ${processes[i]} finished. New Work: [${work.join(', ')}]`);
          found = true;
          count++;
        }
      }
    }

    if (!found) {
      log.push("System is in an UNSAFE state. No process can proceed.");
      return {
        safe: false,
        safe_sequence: [],
        need_matrix: need,
        log
      };
    }
  }

  log.push("System is in a SAFE state.");
  return {
    safe: true,
    safe_sequence: safeSequence,
    need_matrix: need,
    log
  };
}

/**
 * Deadlock Detection using DFS for cycle detection in RAG
 */
export function detectDeadlock(input: DeadlockInput) {
  const { nodes, edges } = input;
  const adj: Record<string, string[]> = {};
  nodes.forEach(node => adj[node.id] = []);
  edges.forEach(edge => adj[edge.source].push(edge.target));

  const visited = new Set<string>();
  const recStack = new Set<string>();
  let cyclePath: string[] = [];

  function findCycle(u: string, path: string[]): boolean {
    visited.add(u);
    recStack.add(u);
    path.push(u);

    for (const v of adj[u]) {
      if (!visited.has(v)) {
        if (findCycle(v, path)) return true;
      } else if (recStack.has(v)) {
        // Cycle found
        const cycleStartIdx = path.indexOf(v);
        cyclePath = path.slice(cycleStartIdx);
        return true;
      }
    }

    recStack.delete(u);
    path.pop();
    return false;
  }

  let deadlock = false;
  for (const node of nodes) {
    if (!visited.has(node.id)) {
      if (findCycle(node.id, [])) {
        deadlock = true;
        break;
      }
    }
  }

  const processesInvolved = cyclePath.filter(id => {
    const node = nodes.find(n => n.id === id);
    return node?.type === 'process';
  });

  return {
    deadlock,
    processes_involved: processesInvolved,
    cycle_path: cyclePath,
    graph_data: { nodes, edges }
  };
}

/**
 * Deadlock Recovery implementation
 */
export function recoverDeadlock(input: RecoveryInput) {
  const { allocation, available, deadlockedProcesses, strategy } = input;
  const steps: string[] = [];
  let currentAllocation = allocation.map(row => [...row]);
  let currentAvailable = [...available];

  if (strategy === 'terminate-all') {
    steps.push(`Terminating all deadlocked processes: ${deadlockedProcesses.join(', ')}`);
    deadlockedProcesses.forEach(pId => {
      const pIdx = parseInt(pId.replace('P', ''));
      for (let j = 0; j < currentAvailable.length; j++) {
        currentAvailable[j] += currentAllocation[pIdx][j];
        currentAllocation[pIdx][j] = 0;
      }
      steps.push(`Process ${pId} terminated. Resources released.`);
    });
  } else if (strategy === 'terminate-one-by-one') {
    // Simple heuristic: terminate process with most resources first to recover faster
    const sorted = [...deadlockedProcesses].sort((a, b) => {
      const idxA = parseInt(a.replace('P', ''));
      const idxB = parseInt(b.replace('P', ''));
      const sumA = currentAllocation[idxA].reduce((s, v) => s + v, 0);
      const sumB = currentAllocation[idxB].reduce((s, v) => s + v, 0);
      return sumB - sumA;
    });

    for (const pId of sorted) {
      steps.push(`Terminating process ${pId} (one-by-one strategy).`);
      const pIdx = parseInt(pId.replace('P', ''));
      for (let j = 0; j < currentAvailable.length; j++) {
        currentAvailable[j] += currentAllocation[pIdx][j];
        currentAllocation[pIdx][j] = 0;
      }
      steps.push(`Process ${pId} terminated. Available resources: [${currentAvailable.join(', ')}]`);
    }
  } else if (strategy === 'preemption') {
    steps.push(`Attempting resource preemption from deadlocked processes: ${deadlockedProcesses.join(', ')}`);
    // Preempt resources one process at a time, releasing their held resources
    const remaining = [...deadlockedProcesses];
    while (remaining.length > 0) {
      // Pick process with fewest total resources (least disruption)
      remaining.sort((a, b) => {
        const idxA = parseInt(a.replace('P', ''));
        const idxB = parseInt(b.replace('P', ''));
        const sumA = currentAllocation[idxA].reduce((s, v) => s + v, 0);
        const sumB = currentAllocation[idxB].reduce((s, v) => s + v, 0);
        return sumA - sumB;
      });
      const victim = remaining.shift()!;
      const victimIdx = parseInt(victim.replace('P', ''));
      const preempted = [...currentAllocation[victimIdx]];
      for (let j = 0; j < currentAvailable.length; j++) {
        currentAvailable[j] += currentAllocation[victimIdx][j];
        currentAllocation[victimIdx][j] = 0;
      }
      steps.push(`Preempted resources [${preempted.join(', ')}] from process ${victim}. Available: [${currentAvailable.join(', ')}]`);
      steps.push(`Process ${victim} rolled back to a safe checkpoint.`);

      // Check if remaining processes can now proceed
      const canProceed = remaining.every(pId => {
        const pIdx = parseInt(pId.replace('P', ''));
        return currentAllocation[pIdx].every((val, j) => val <= currentAvailable[j]);
      });
      if (canProceed && remaining.length > 0) {
        steps.push(`Deadlock resolved! Remaining processes [${remaining.join(', ')}] can now proceed.`);
        break;
      }
    }
  }

  return {
    strategy_used: strategy,
    recovery_steps: steps,
    updated_allocation: currentAllocation,
    system_safe: true
  };
}

/**
 * Simulation Engine — runs a step-by-step resource allocation scenario
 * using the Banker's Algorithm to check safety at each step.
 */
export function runSimulation(input: any) {
  const numProcesses = input.numProcesses || 4;
  const numResources = input.numResources || 3;
  const log: string[] = [];

  // Generate initial system state
  const totalResources = Array.from({ length: numResources }, () => Math.floor(Math.random() * 8) + 4);
  const allocation: number[][] = Array.from({ length: numProcesses }, () =>
    Array.from({ length: numResources }, () => 0)
  );
  const max: number[][] = Array.from({ length: numProcesses }, () =>
    totalResources.map(t => Math.floor(Math.random() * (t + 1)))
  );
  const available = [...totalResources];
  const processes = Array.from({ length: numProcesses }, (_, i) => `P${i}`);
  const resources = Array.from({ length: numResources }, (_, i) => String.fromCharCode(65 + i));

  log.push(`Starting simulation with ${numProcesses} processes and ${numResources} resources.`);
  log.push(`Total Resources: [${totalResources.join(', ')}]`);
  log.push(`Available: [${available.join(', ')}]`);

  // Simulate allocation steps
  let deadlockDetected = false;
  const steps = numProcesses * 2;
  for (let step = 0; step < steps; step++) {
    const pIdx = Math.floor(Math.random() * numProcesses);
    const rIdx = Math.floor(Math.random() * numResources);
    const need = max[pIdx][rIdx] - allocation[pIdx][rIdx];

    if (need > 0 && available[rIdx] > 0) {
      const amount = Math.min(need, available[rIdx], Math.ceil(Math.random() * 2));
      // Tentatively allocate
      allocation[pIdx][rIdx] += amount;
      available[rIdx] -= amount;

      // Check safety
      const safetyResult = bankersSafetyCheck({ processes, resources, allocation, max, available });
      if (safetyResult.safe) {
        log.push(`[Step ${step + 1}] Granted ${amount} unit(s) of ${resources[rIdx]} to ${processes[pIdx]}. System SAFE. Sequence: ${safetyResult.safe_sequence.join(' → ')}`);
      } else {
        // Rollback
        allocation[pIdx][rIdx] -= amount;
        available[rIdx] += amount;
        log.push(`[Step ${step + 1}] Denied ${amount} unit(s) of ${resources[rIdx]} to ${processes[pIdx]}. Would cause UNSAFE state — request blocked.`);
      }
    } else {
      log.push(`[Step ${step + 1}] ${processes[pIdx]} has no pending need for ${resources[rIdx]}, skipping.`);
    }
  }

  // Final check
  const finalResult = bankersSafetyCheck({ processes, resources, allocation, max, available });
  const status = finalResult.safe ? 'safe' : 'unsafe';
  log.push(`--- Simulation Complete ---`);
  log.push(`Final Available: [${available.join(', ')}]`);
  log.push(`Final Status: ${status.toUpperCase()}`);
  if (finalResult.safe) {
    log.push(`Safe Sequence: ${finalResult.safe_sequence.join(' → ')}`);
  }

  return {
    status,
    running_processes: processes.filter((_, i) => allocation[i].some(v => v > 0)),
    blocked_processes: finalResult.safe ? [] : processes,
    analysis_log: log
  };
}

File: algorithms.py
"""
Core deadlock detection and prevention algorithms.
These functions are UI-independent and can be used standalone.
"""


def bankers_safety_check(processes, resources, allocation, max_matrix, available):
    """
    Banker's Algorithm — Safety Check.
    
    Args:
        processes: List of process IDs (e.g., ['P0', 'P1', 'P2'])
        resources: List of resource types (e.g., ['A', 'B', 'C'])
        allocation: 2D list of currently allocated resources [process][resource]
        max_matrix: 2D list of maximum needed resources [process][resource]
        available: List of available resources [resource]
    
    Returns:
        (safe, safe_sequence, need_matrix, log)
        - safe: Boolean, True if system is in safe state
        - safe_sequence: List of processes in safe order (empty if unsafe)
        - need_matrix: 2D list of additional resources needed [process][resource]
        - log: List of log messages describing execution
    """
    n = len(processes)
    m = len(resources)

    # Calculate Need matrix
    need = [[max_matrix[i][j] - allocation[i][j] for j in range(m)] for i in range(n)]
    
    work = available[:]
    finish = [False] * n
    safe_sequence = []
    log = []

    log.append(f"Initial Work (Available): {work}")

    count = 0
    while count < n:
        found = False
        for i in range(n):
            if not finish[i]:
                # Check if process i's needs can be satisfied
                if all(need[i][j] <= work[j] for j in range(m)):
                    log.append(f"Process {processes[i]} can be allocated. Need: {need[i]} <= Work: {work}")
                    # Process i can finish; release its resources
                    work = [work[j] + allocation[i][j] for j in range(m)]
                    finish[i] = True
                    safe_sequence.append(processes[i])
                    log.append(f"Process {processes[i]} finished. New Work: {work}")
                    found = True
                    count += 1
        
        if not found:
            log.append("System is in an UNSAFE state. No process can proceed.")
            return False, [], need, log

    log.append("System is in a SAFE state.")
    return True, safe_sequence, need, log


def detect_deadlock(nodes, edges):
    """
    Detect deadlock using DFS-based cycle detection in Resource Allocation Graph.
    
    Args:
        nodes: List of node dicts with 'id' and 'type' (e.g., {'id': 'P1', 'type': 'process'})
        edges: List of edge dicts with 'source' and 'target' (e.g., {'source': 'P1', 'target': 'R1'})
    
    Returns:
        (deadlock, processes_involved, cycle_path)
        - deadlock: Boolean, True if cycle found
        - processes_involved: List of process IDs in the cycle
        - cycle_path: List of node IDs forming the cycle
    """
    # Build adjacency list
    adj = {n['id']: [] for n in nodes}
    for e in edges:
        adj[e['source']].append(e['target'])

    visited = set()
    rec_stack = set()
    cycle_path = []

    def find_cycle(u, path):
        """DFS to find a cycle. Returns True if cycle found and sets cycle_path."""
        nonlocal cycle_path
        visited.add(u)
        rec_stack.add(u)
        path.append(u)
        
        for v in adj[u]:
            if v not in visited:
                if find_cycle(v, path):
                    return True
            elif v in rec_stack:
                # Cycle found: from current node back to v
                idx = path.index(v)
                cycle_path = path[idx:]
                return True
        
        rec_stack.discard(u)
        path.pop()
        return False

    deadlock = False
    for node in nodes:
        if node['id'] not in visited:
            if find_cycle(node['id'], []):
                deadlock = True
                break

    # Extract process IDs from cycle path
    processes_involved = [nid for nid in cycle_path 
                         if any(n['id'] == nid and n['type'] == 'process' for n in nodes)]

    return deadlock, processes_involved, cycle_path


def recover_deadlock(allocation, available, deadlocked_processes, strategy):
    """
    Deadlock Recovery with 3 strategies.
    
    Args:
        allocation: 2D list of resource allocations [process_index][resource_index]
        available: List of available resources [resource_index]
        deadlocked_processes: List of process IDs in deadlock (e.g., ['P0', 'P1'])
        strategy: One of 'terminate-all', 'terminate-one-by-one', or 'preemption'
    
    Returns:
        (strategy_used, steps, updated_allocation)
        - strategy_used: The strategy selected
        - steps: List of step descriptions
        - updated_allocation: Updated allocation matrix
    """
    steps = []
    current_alloc = [row[:] for row in allocation]
    current_avail = available[:]
    num_resources = len(available)

    if strategy == 'terminate-all':
        steps.append(f"Terminating all deadlocked processes: {', '.join(deadlocked_processes)}")
        for pid in deadlocked_processes:
            idx = int(pid.replace('P', ''))
            for j in range(num_resources):
                current_avail[j] += current_alloc[idx][j]
                current_alloc[idx][j] = 0
            steps.append(f"Process {pid} terminated. Resources released.")

    elif strategy == 'terminate-one-by-one':
        # Terminate processes in order of resource usage (highest first)
        sorted_procs = sorted(deadlocked_processes,
                              key=lambda p: sum(current_alloc[int(p.replace('P', ''))]),
                              reverse=True)
        for pid in sorted_procs:
            steps.append(f"Terminating process {pid} (one-by-one strategy).")
            idx = int(pid.replace('P', ''))
            for j in range(num_resources):
                current_avail[j] += current_alloc[idx][j]
                current_alloc[idx][j] = 0
            steps.append(f"Process {pid} terminated. Available resources: {current_avail}")

    elif strategy == 'preemption':
        steps.append(f"Attempting resource preemption from: {', '.join(deadlocked_processes)}")
        remaining = list(deadlocked_processes)
        
        while remaining:
            remaining.sort(key=lambda p: sum(current_alloc[int(p.replace('P', ''))]))
            victim = remaining.pop(0)
            victim_idx = int(victim.replace('P', ''))
            preempted = current_alloc[victim_idx][:]
            
            for j in range(num_resources):
                current_avail[j] += current_alloc[victim_idx][j]
                current_alloc[victim_idx][j] = 0
            
            steps.append(f"Preempted resources {preempted} from {victim}. Available: {current_avail}")
            steps.append(f"Process {victim} rolled back to a safe checkpoint.")
            
            if remaining:
                can_proceed = all(
                    all(current_alloc[int(p.replace('P', ''))][j] <= current_avail[j] 
                        for j in range(num_resources))
                    for p in remaining
                )
                if can_proceed:
                    steps.append(f"Deadlock resolved! Remaining processes [{', '.join(remaining)}] can now proceed.")
                    break

    return strategy, steps, current_alloc
